### DISCUSSION TITLE
Container sizes, Names, and consistency in UI

### DISCUSSION THREAD

**Discussion by Saevon** [FakeFishGames/Barotrauma#13114] [Ideas and suggestions]
Currently there are a bunch of nice container sprites, however each has individually unique limitations. We should be able to customize a bunch of properties!

This way if you make use a "Medical compartment" resizing it to look larger, and adding sprite Doodads, it could have 6 slots, and be named "Wiring Compartment".   Or the basic "Supplies Cabinet" could be renamed to "Oxygen Mask Container" or "Welding Container".  Or the "Medium Steel Cabinet" could be renamed to "Reactor Storage" or "Engine Storage"

**(Suggestion 1)** Containers should be customizeble allowing:
- `ContainerSize` (number of slots)
  * Anything with Sprites (weapon containers, shelves crate storage) would have this disabled
- `ContainerName (translated)` allowing you to specify what its for when a user hovers
  * (see way below for suggestions) By default will be the name of the doodad like "Medical Compartment"
- `Authorization` (rename Picked Requiring, now on ALL container types)
  - This should be editable in game. Allowing you to add a card to it IF `Allow In-Game Editing` is enabled (off by default for containers) requires the current Authorization to edit
- `Containable Restrictions`
- `Autofill Categories` (Split out from the tags section)

(The moves and renames help make this more intuitive to sub builders)

**(Suggestion 2)** The following deco items should also be containers already (By default non-interactive? aka not act like containers, but usable as containers when enabled)
- Box1-5
- Boxes, Metal Box
- Cabinets
- Clothes Rack
- Engineer Equipment
- Large Shelf
- Freezer
- Frozen Seed Display Shelf, Seed Display
- Medical Rack
- Weapons display
- Shelf  (all the deco ones without a unique name)
- Item Display Shelf
- Produce Boxes

**(Suggestion 3)** All container types should be a category in the Sub Builder with items appearing in both their original category and this category.

Example:  Ammunition shelf, Secure container, Weapon Display cases should ALL be both under weapons AND under Containers
Example:  Engineer Equipment should be under Deco AND containers

Or alternatively they all need to be differently searchable. Note the same thing applies to other items which are hard to find  (And a few you already have grouped!)…
- Lights (+spotlights)
- Containers
- Back Walls  (Back Wall, Background)
- Walls (Interiors with Health)
- Walkables (Walls, Platforms, Ladders, Stairs, Spawnpoint, Waypoint)
- Doors & Hatches (+Ducts, Gaps, Hulls?)
- Decals  (Fliers, Paintings, Grafitti, Decals, Signs, Labels, Posters, Pictures, Plaque, Blueprint, Banner, Boards, Target, Advertisement, Labels)
  * Note:  Seperatist decoration should all be named after the deco type: "Poster, Grafitti, Boxes, Machinery"
  * Note: Bandit Prop should all be named after the deco type: "Poster, Grafitti, Boxes, Machinery"
- Displays (Monitors, Lightboxes, TVs, Scree)

# Container Names

Here is a list of some common container names (Will maintain based on suggestions) which could be added to translations:

- Reactor
- Wiring
- Electrical
- Oxygen
- Minerals
- Medical
- Weapons
- Armoury
- Diving Gear
- Ammunition
- Tools
- Engineering
- Clothing
- Seeds
- Toxic
- Instruments

- Loose Panel
- Contraband

**Comment by SomeRandomNoobKekeke** [comment]
Couldn't you just make a mod with a pack of these containers with predefined sizes, names and restrictions?
![barodev small](https://github.com/Regalis11/Barotrauma/assets/122838333/971accb8-d44e-4c86-9650-3c90af22e42f)

**Reply by Saevon** [reply]
sooooo  we'd need one item for every pair of "name" to "sprite",,, I wonder how my exponential functions are 🤔

**Comment by Arctodites** [comment]
I do agree that we need more flexible and varied options for storage cabinetry.  For one thing, all of the purely cosmetic display shelves and display cases should be interactable, and able to hold (< display) at least one item.

But turning all of the cosmetic sprites into actual containers would balloon the number of storage slots for items on the sub and potentially unbalance a lot of things.

If we simply had a few generic container types with fixed (1, 2, 3, 6 slot) capacities, e.g.:

- 'shelf,'
- 'chest,'
- 'box,'
- 'cabinet,'
- 'locker'
- 'drawer,'
- 'rack,'
- 'bin'

with generic and easily resizable / recolourable spritework, it would be easy to adapt them to most needs.  The word 'compartment' (as in 'medical compartment') should be avoided in future, though, because a 'compartment' is the term for a room on a submarine.

On the other hand, having more differentiated storage options (both typewise and programmatically) could potentially make it easier in the future for the devs to improve the coding for how AI bots clean up and store items.  And making containers that are visually distinct in their iconography, their spritework, and which can be labelled or recoloured in-game would help to prevent players from just randomly sticking items in the nearest locker because they are uncertain where those items properly belong.  Typically, I use the ItemMsgInteractSelect field of most interactable items as a kind of makeshift custom label, because anyone who plays Barotrauma already knows that default-clicking on an interactable object will interact with that object in a default manner, but really it would be better if ALL interactable objects in the game had a 'NAME' field that would display custom text beneath (or instead of) the generic name when hovered-over in-game.

Adding to your points above, a few other things I'd love to see for containers:
- items contained in any 'loose panel' should not be locatable using the status monitor's item finder, nor should they be sellable
- there should be radio button options in the sub editor to tag ANY container (including wearables like toolbelt, backpack, etc.) with the _dontsellitems_, _donttakeitems_, and _hidden_ tags, and for those tags to work as expected (i.e. prevent sale of contents, prevent bots from taking contents, prevent bots from using container).  Last I checked, these tags do not work on certain items (e.g. toolbelt).

**Reply by Saevon** [reply]
> But turning all of the cosmetic sprites into actual containers would balloon the number of storage slots for items on the sub and potentially unbalance a lot of things.

Considering `ContainerSize` would be an added field, then any existing storage would have their CURRENT value added (0 for most cosmetic items).  But newly placed cosmetics might have a different value after this change (some might still be better defaulted to 0)

Thats why I specified "by default non interactive"!

> items contained in any 'loose panel' should not be locatable using the status monitor's item finder, nor should they be sellable

agreed!!

> there should be radio button options in the sub editor to tag ANY container (including wearables like toolbelt, backpack, etc.) with the dontsellitems, donttakeitems, and hidden tags, and for those tags to work as expected (i.e. prevent sale of contents, prevent bots from taking contents, prevent bots from using container). Last I checked, these tags do not work on certain items (e.g. toolbelt).

All the hidden tags and features that seem to get passed around like rumours annoy me a lot in this game… There needs to be way more information available around all this. E.g. the tags field could have a button that lists all existing special tags, with descriptions.

**Comment by Ek-Videogames** [comment]
Agreed that more translation strings for generic containers would be good, 

additionally with recent changes to item stacks it really feels like there needs to be a 'medium-small' container size which holds more than the pitiful 6 items the large 'medical-compartment' & supply cabinet holds - but fewer than a small cabinet or secure cabinet, something around 10 or 12 items maybe?

> But turning all of the cosmetic sprites into actual containers would balloon the number of storage slots for items on the sub and potentially unbalance a lot of things.

This already exists, it's called "Make the item non-interactable and then place an invisible large steel cabinet on top of it" 
Making the new containers operable in their own capacity would reduce item count, and better yet if they were customizable as previously stated, so people could make them act as 'outpost cabinets' (6 inventory) or as submarine containers more akin to steel cabinet (20 inventory).

Also, though this is somewhat less related to customization and moreso just optimization: A setting for all sorts of items not only containers which just sets an item as "don't simulate" (IE character interaction, lighting, item-components, status-effects) because we have "hide in game" or "non interactable" or "invulnerable to damage" but those settings don't stop certain items from continuing to have their 'dynamic' behaviors, the reason this would be useful is so people can use certain items as decorative structure elements without causing performance issues due to them being part of the item count.

**Reply by Saevon** [reply]
yeah! or cabinets specifically just for medium items, or any ratio of larger bags to smaller items.
